src/theory/strings/array_core_solver.cpp

```diff
@@ -95,63 +95,68 @@ void ArrayCoreSolver::checkUpdate(const std::vector<Node>& updateTerms)
     // inference rule is:
     // (seq.update x i a) in TERMS
     // (seq.nth t j) in TERMS
-    // t == (seq.update x i a)
-    // ----------------------------------------------------------------------
+    // t = (seq.update x i a)
+    // ---------------------------------------------------------------------
     // (seq.nth (seq.update x i a) j) =
     //   (ITE, j in range(i, i+len(a)), (seq.nth a (j - i)),  (seq.nth x j))
 
-    // t == (seq.update x i a) =>
-    // (seq.nth t j) = (ITE, j in range(i, i+len(a)), (seq.nth a (j - i)),
-    // (seq.nth x j))
-
     // note that the term could rewrites to a skolem
     // get proxy variable for the update term as t
     Node termProxy = d_termReg.getProxyVariableFor(n);
     Trace("seq-update") << "- " << termProxy << " = " << n << std::endl;
     std::vector<Node> exp;
     d_im.addToExplanation(termProxy, n, exp);
 
-    // optimization: add a short cut t == (seq.update n[0] n[1] n[2]) => t[i] ==
-    // n[2][0]
+    // optimization: add a short cut
+    // t = (seq.update n[0] n[1] n[2])
+    // ---------------------------------------------------------------------
+    // nth(t, n[1]) = (ITE, len(n[0]) > 0, nth(n[2], 0), nth(n[0], n[1]))
     Node left = nm->mkNode(SEQ_NTH, termProxy, n[1]);
-    Node right =
-        nm->mkNode(SEQ_NTH, n[2], nm->mkConstInt(Rational(0)));  // n[2][0]
+    Node cond = nm->mkNode(
+        GT, nm->mkNode(STRING_LENGTH, n[0]), nm->mkConstInt(Rational(0)));
+    Node body1 = nm->mkNode(SEQ_NTH, n[2], nm->mkConstInt(Rational(0)));
+    Node body2 = nm->mkNode(SEQ_NTH, n[0], nm->mkConstInt(Rational(0)));
+    Node right = nm->mkNode(ITE, cond, body1, body2);
     Node lem = nm->mkNode(EQUAL, left, right);
-    Trace("seq-array-debug") << "enter" << std::endl;
-    sendInference(exp, lem, InferenceId::STRINGS_ARRAY_NTH_UPDATE);
+    sendInference(exp, lem, InferenceId::STRINGS_ARRAY_NTH_UPDATE_SHORTCUT1);
 
     // enumerate possible index
-    for (auto nth : d_index_map)
+    for (const auto& nth : d_index_map)
     {
       Node seq = nth.first;
       if (d_state.areEqual(seq, n) || d_state.areEqual(seq, n[0]))
       {
-        std::set<Node> indexes = nth.second;
+        const std::set<Node>& indexes = nth.second;
         for (Node j : indexes)
         {
-          // optimization: add a short cut for special case (seq.update n[0]
-          // n[1] (seq.unit e))
+          // optimization: add a short cut for special case
+          // n[2] is a unit term
+          // n[1] != j
+          // ---------------------------------------------------------------
+          // nth(t, j) = nth(n[0], j)
           if (n[2].getKind() == SEQ_UNIT)
           {
             left = nm->mkNode(DISTINCT, n[1], j);
             Node nth1 = nm->mkNode(SEQ_NTH, termProxy, j);
             Node nth2 = nm->mkNode(SEQ_NTH, n[0], j);
             right = nm->mkNode(EQUAL, nth1, nth2);
             lem = nm->mkNode(IMPLIES, left, right);
```

Comment:
We should think about whether we really want an implication here or whether we want to check the current state and only emit the lemma (with the left-hand side of the implication as the explanation) when the left-hand side is full-filled.